torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,160 @@ namespace rpc {
 
 constexpr size_t WorkerInfo::MAX_NAME_LEN;
 
+// kRpcRetryDuration is the initial duration between consecutive RPC retries.
+constexpr auto kRpcRetryDuration = std::chrono::milliseconds(1000);
+// kBackoff is the multiplier for the exponential backoff algorithm.
+constexpr float kBackoff = 1.5;
+
 RpcAgent::RpcAgent(
     WorkerInfo workerId,
     std::unique_ptr<RequestCallback> cb,
     std::chrono::milliseconds rpcTimeout)
     : workerInfo_(std::move(workerId)),
       cb_(std::move(cb)),
       rpcTimeout_(rpcTimeout),
-      profilingEnabled_(false) {}
+      profilingEnabled_(false) {
+  rpcAgentRunning_.store(true);
+  rpcRetryThread_ = std::thread(&RpcAgent::retryExpiredRpcs, this);
+}
+
+RpcAgent::~RpcAgent() {
+  if (rpcAgentRunning_.load()) {
+    shutdown();
+  }
+}
+
+void RpcAgent::shutdown() {
+  if (!rpcAgentRunning_.exchange(false)) {
+    return;
+  }
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
+  shutdownImpl();
+}
+
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
```

Comment:
Just realized that in the code this is used as maxTries and not maxRetries. For example if maxRetries is 3, we should try a total of 4 RPCs (the original RPC and then 3 retries). Can we update the code to follow maxRetries appropriately?